powershap/shap_wrappers/shap_explainer.py

```diff
@@ -96,23 +98,21 @@ def explain(
         random_col_name = "random_uniform_feature"
         assert random_col_name not in X.columns
 
-        X = X.copy(deep=True)
-
```

Comment:
In hindsight (as written in the general code review comment), the most elegant solution to me is adding the random col after the train & test split - https://github.com/predict-idlab/powershap/pull/38#discussion_r1189464803

Reply:
I thought of this "mutating the data" possibility too, yeah. At first, I was nervous to remove these copies, so I had to do a lot of digging around to prove to myself that it was safe. Here is my current impression of how the code works, as currently written.

In `powershap.py`, there is this code:

        X, y = self._explainer._validate_data(self._validate_data, X, y, multi_output=True)

        self._print("Starting powershap")

        X = pd.DataFrame(data=X, columns=list(range(X.shape[1])))`

The first line of this code has the following effects, from what I can tell by reading the _validate_data and _check_X_y functions in sklearn:

1. If the input X is a numpy array and isn't changed by the validation in a way that would create a copy, it may return a reference to itself. (No copy yet, in this case.)
2. If the input X is a numpy array and is changed by the validation in certain ways, it may cause X to be replaced with a copy of the input array. (One copy made here.)
3. If the input X is not a numpy array (e.g. it's a Pandas DataFrame), it will be converted to a numpy array by the validation routine. (This creates a new copy.)

In all of the above cases, the following pd.DataFrame() call will convert the resulting array to a pandas dataframe, which creates a new copy of the data. So, the result of these two lines of code is that either one or two copies are made.

In any of these three cases, PowerShap is no longer operating on the original data by the time the fit operation runs... it's operating on a copy of the data. So I think if fit gets cancelled in the middle of doing its work, the original data won't get mutated.

I tested my assumptions here by printing the memory references in the case where a numpy array is passed to powershap.fit:

    before powershap.fit: 139661866098288
    X inside fit at the beginning: 139661866098288
    after validate: 139661866098288
    after pd.DataFrame: 139661217669808

And in the case where a pandas dataframe is passed in:

    before powershap.fit: 140176839600976
    X inside fit at the beginning: 140176839600976
    after validate: 140175763418832
    after pd.DataFrame: 140175763333072

So as you can see, by the time the pd.DataFrame call ends, fit is always operating on a copy of the data. The only thing to worry about, I think, is if these lines change in the future in a way that doesn't ensure that a copy of the data is made.

Reply:
Very clear explanation @lorenjan 

It would be great if you could add a condensed version of the above explanation to the code base ([here](https://github.com/predict-idlab/powershap/blob/6b54fb728947b76ed4b1a6b855fcaaf5b1514d64/powershap/powershap.py#L366-L370)) - as such we will be aware of this behavior if we ever intend to update the code in the future.

Reply:
Will do!

Reply:
Awesome!! Kudos for the tests :fire: